pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,429 @@
+var _ reconcile.Reconciler = &ReconcileClusterDeployment{}
+
+// ReconcileClusterDeployment reconciles a ClusterDeployment object
+type ReconcileClusterDeployment struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ClusterDeployment object and makes changes based on the state read
+// and what is in the ClusterDeployment.Spec
+// TODO(Raf) Add finalizers and clean up
+func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ClusterDeployment")
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1alpha1.ClusterDeployment{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Do not make do anything if the cluster is not a GCP cluster.
+	val, ok := cd.Labels[clusterPlatformLabel]
+	if !ok || val != clusterPlatformGCP {
+		reqLogger.Info("not a gcp cluster")
+		return reconcile.Result{}, nil
+	}
+
+	// Do not do anything if the cluster is not a Red Hat managed cluster.
+	val, ok = cd.Labels[clusterDeploymentManagedLabel]
+	if !ok || val != "true" {
+		reqLogger.Info("not a managed cluster")
+		return reconcile.Result{}, nil
+	}
+
+	//Do not reconcile if cluster is installed or remove cleanup and remove project
+	if cd.Spec.Installed {
+		// TODO(Raf) Cleanup and remove project if being deleted once Hive is finished uninstalling
+		reqLogger.Info(fmt.Sprintf("cluster %v is in installed state", cd.Name))
+		return reconcile.Result{}, nil
+	}
+
+	// Get org creds from secret
+	creds, err := GetOrgGCPCreds(r.client, operatorNamespace)
+	if err != nil {
+		reqLogger.Error(err, "Could not get org Creds from secret", "Secret Name", orgGcpSecretName, "Operator Namespace", operatorNamespace)
+		return reconcile.Result{}, err
+	}
+
+	// Check if gcpSecretName in cd.Namesapce exists we are done
+	if SecretExists(r.client, gcpSecretName, cd.Namespace){
+		reqLogger.Info(fmt.Sprintf("Secret: %s already exists in Namespace: %s :: Nothing to do", gcpSecretName, cd.Namespace))
+		return reconcile.Result{}, nil
+	}
+
+	// Skip code block to create project for now until we have permissions to test
+	if false {
+		// Check if platform projectID string exists.
+		if cd.Spec.Platform.GCP.ProjectID != "" && cd.Spec.Platform.GCP.Region != "" {
```

Comment:
Is the project ID something from GCP? How would the cluster deployment have it before this operator created the project?

Reply:
It would need to come from OCM as randomly generated value, similar to namespace in Kubernetes.